### PR TITLE
uuid: Add validation for uuid which has hyphen after any group of four hexdigits

### PR DESCRIPTION
**Pull request by stuckinforloop**
This fixes issue: #60011

Release note (bug fix): Fixed a bug. Postgres allows a UUID string in which hyphens were placed after a group of 4 hexdigits. This was not implemented in CockroachDB. This PR solves this issue.

### CODE REVIEW

**Review by rafiss — CHANGES_REQUESTED** [review]
this is off to a good start! i left a comment; we actually want to allow even more inputs

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63157#-:-MXom6wM0KXFPZlF5pNu:b-9gihnc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @neel229)

___
*[pkg/util/uuid/codec.go, line 146 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/63157#-MXolIF08b8monAUh1Z7:-MXolIF08b8monAUh1Z8:b-jh82mc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c5f8b4d79d8d0c2956ca231b3606f3bbb41935e/pkg/util/uuid/codec.go#L146)):*
so we'd actually want to be a little bit more permissive

the postgres docs say that it should support 
> adding a hyphen after any group of four digits.

https://www.postgresql.org/docs/current/datatype-uuid.html

so basically, after every 4 digits, there may or may not be a hyphen.

these UUIDs (and others) should all count as valid after this change:

```
6ba7-b8109dad11d180b400c04fd4-30c8
{6ba7-b8109dad11d180b400c04fd4-30c8}
urn:uuid:6ba7-b8109dad11d180b400c04fd4-30c8
6ba7-b810-9dad11d180b400c04fd4-30c8
6ba7-b810-9dad-11d180b400c04fd4-30c8
```

but these ones are invalid:
```
6ba7b810-9dad11d180b400c04fd430-c8
6-ba7b810-9dad11d180b400c04fd430c8
6ba7b810-9dad11d180b400c04fd430c8-
-6ba7b810-9dad11d180b400c04fd430c8
6ba7-b810-9dad--11d1-80b4-00c0-4fd4-30c8
6ba7-b810-9dad-11d1-80b4-00c0-4fd4-30c8--
6ba7-b810--9dad-11d180b4-00c04fd4-30c8-
```

**Review by stuckinforloop — COMMENTED** [review]
Okay, I will start working on allowing more inputs.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63157#-:-MXovuFX635AZ5_q20n3:b-abgnal)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @neel229)

**Review by stuckinforloop — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63157)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/util/uuid/codec.go, line 146 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/63157#-MXolIF08b8monAUh1Z7:-MXowAGG2GVoykqqxvF7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c5f8b4d79d8d0c2956ca231b3606f3bbb41935e/pkg/util/uuid/codec.go#L146)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

so we'd actually want to be a little bit more permissive

the postgres docs say that it should support 
> adding a hyphen after any group of four digits.

https://www.postgresql.org/docs/current/datatype-uuid.html

so basically, after every 4 digits, there may or may not be a hyphen.

these UUIDs (and others) should all count as valid after this change:

```
6ba7-b8109dad11d180b400c04fd4-30c8
{6ba7-b8109dad11d180b400c04fd4-30c8}
urn:uuid:6ba7-b8109dad11d180b400c04fd4-30c8
6ba7-b810-9dad11d180b400c04fd4-30c8
6ba7-b810-9dad-11d180b400c04fd4-30c8
```

but these ones are invalid:
```
6ba7b810-9dad11d180b400c04fd430-c8
6-ba7b810-9dad11d180b400c04fd430c8
6ba7b810-9dad11d180b400c04fd430c8-
-6ba7b810-9dad11d180b400c04fd430c8
6ba7-b810-9dad--11d1-80b4-00c0-4fd4-30c8
6ba7-b810-9dad-11d1-80b4-00c0-4fd4-30c8--
6ba7-b810--9dad-11d180b4-00c04fd4-30c8-
```
</blockquote></details>

On it.
